### PR TITLE
autoinstall: fix interaction with early-commands 

### PR DESCRIPTION
**Pull request by dbungert**
Early commands is currently documented as follows:
> The autoinstall config is available at /autoinstall.yaml
> (irrespective of how it was provided) and the file will be re-read
> after the early-commands have run to allow them to alter the config
> if necessary.

The previous change to let cloud take precedence over iso location broke
this functionality.  Fix that by copying to the iso location the file of
choice.

Continuing from #1237

### CODE REVIEW

File: subiquitycore/file_util.py [hunk 1]
```diff
@@ -73,3 +74,16 @@ def generate_config_yaml(filename, content, **kwargs):
         now = datetime.datetime.utcnow()
         tf.write(f'# Autogenerated by Subiquity: {now} UTC\n')
         tf.write(yaml.dump(content))
+
+
+def copy_file_if_exists(source, target):
+    """If source exists, copy to destination.  Ignore error that dest may be a
+    duplicate.  Create destination parent dirs as needed."""
+    if source is None or not os.path.exists(source):
```
Comment (ogayot) [inline subiquitycore/file_util.py:82]: What would you think of marking `source` as a `str` and dropping the check for `source is None`?
I understand that it means making the call to `copy_file_if_exists` conditional (i.e., based on whether `self.autoinstall is None`) but I don't know about silently ignoring `None` values in this function.

**Review by ogayot — APPROVED** [review]
Would a symlink be an option rather than copying the contents of the file?

**Comment by dbungert** [comment]
> Would a symlink be an option rather than copying the contents of the file?

Maybe.  I'm learning about the behavior of autoinstall during a refresh.  It would be strange if early-commands ran twice and double-modified the file, but double-running early commands is also probably bad for many other reasons.

File: subiquitycore/file_util.py (hunk 1 above)
Comment (dbungert) [reply subiquitycore/file_util.py:82]: I'll do this if we don't end up with the symlink

Comment (ogayot) [reply subiquitycore/file_util.py:82]: Alright, choice is up to you!

**Comment by mwhudson** [comment]
> > Would a symlink be an option rather than copying the contents of the file?

I worry a little about early-commands editing the file in a way that does not respect symlinks. E.g. one that does "cp /cdrom/my-autoinstall.yaml /autoinstall.yaml" or similar.

> Maybe. I'm learning about the behavior of autoinstall during a refresh. It would be strange if early-commands ran twice and double-modified the file, but double-running early commands is also probably bad for many other reasons.

early-commands should definitely only run once!

File: subiquity/server/tests/test_server.py [hunk 2]
```diff
@@ -40,55 +42,73 @@ def setUp(self):
...
-            self.assertEqual(expected, fp.read())
+        self.assertIsNone(self.server.select_autoinstall())
 
     def test_bogus_autoinstall_argument(self):
         self.server.opts.autoinstall = self.path('nonexistant.yaml')
         with self.assertRaises(Exception):
-            self.server.select_autoinstall_location()
+            self.server.select_autoinstall()
+
+    def test_early_commands_changes_autoinstall(self):
```
Comment (mwhudson) [inline subiquity/server/tests/test_server.py:94]: Could you add a test or modify this test to simulate the autoinstall config being supplied via cloud-init config?

File: subiquity/server/server.py [hunk 3]
```diff
@@ -539,7 +541,7 @@ async def wait_for_cloudinit(self):
                 "cloud-init status: %r, assumed disabled",
                 status_txt)
 
-    def select_autoinstall_location(self):
+    def select_autoinstall(self):
```
Comment (mwhudson) [inline subiquity/server/server.py:544]: I'm not sure this method name is quite appropriate any more. But I don't have any better ideas.

**Review by mwhudson — APPROVED** [review]
I'd like one more test please, as I thought you'd missed a case but on close reading don't think you have.

File: subiquitycore/file_util.py (hunk 1 above)
Comment (dbungert) [reply subiquitycore/file_util.py:82]: Doing the non-symlink implementation, so I have done the above.

File: subiquity/server/server.py (hunk 3 above)
Comment (dbungert) [reply subiquity/server/server.py:544]: I sat here for a few minutes pondering a better name and I have nothing.  If we think of one later I'm open to changing it.

File: subiquity/server/tests/test_server.py (hunk 2 above)
Comment (dbungert) [reply subiquity/server/tests/test_server.py:94]: Good point, it's the entire reason for the PR.  Done!

**Pull request merged**